File: mobile/ios/Runner/BackgroundConversationSyncChannels.swift
import Flutter
import UIKit

final class BackgroundConversationSyncChannels: NSObject, FlutterStreamHandler {
  private static let methodChannelName = "lan_ai_cli_control/background_conversation_sync"
  private static let eventChannelName = "lan_ai_cli_control/background_conversation_sync/events"

  private weak var application: UIApplication?
  private var eventSink: FlutterEventSink?
  private var lastSnapshot: [String: Any?]?
  private var backgroundTask: UIBackgroundTaskIdentifier = .invalid

  init(application: UIApplication = .shared) {
    self.application = application
  }

  func register(binaryMessenger: FlutterBinaryMessenger) {
    let methodChannel = FlutterMethodChannel(
      name: Self.methodChannelName,
      binaryMessenger: binaryMessenger
    )
    methodChannel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result: result)
    }

    let eventChannel = FlutterEventChannel(
      name: Self.eventChannelName,
      binaryMessenger: binaryMessenger
    )
    eventChannel.setStreamHandler(self)
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    if let snapshot = lastSnapshot {
      events(snapshot)
    }
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "isSupported":
      result(true)
    case "start":
      guard let request = request(from: call.arguments as? [String: Any]) else {
        result(FlutterError(
          code: "BAD_ARGUMENT",
          message: "Invalid background conversation sync request",
          details: nil
        ))
        return
      }
      result(startCleanupTask(request: request))
    case "snapshot":
      result(lastSnapshot ?? stoppedSnapshot(message: "iOS background cleanup is inactive"))
    case "stop":
      finishCleanupTask(message: "iOS background cleanup stopped")
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func startCleanupTask(request: BackgroundConversationSyncRequest) -> [String: Any?] {
    endBackgroundTaskIfNeeded()
    backgroundTask = application?.beginBackgroundTask(withName: "ConversationSyncCleanup") { [weak self] in
      self?.finishCleanupTask(message: "iOS background cleanup expired")
    } ?? .invalid

    if backgroundTask == .invalid {
      let snapshot = makeSnapshot(
        status: "failed",
        runningCount: request.runningCount,
        waitingApprovalCount: request.waitingApprovalCount,
        message: "iOS background cleanup was rejected"
      )
      emit(snapshot)
      return snapshot
    }

    let snapshot = makeSnapshot(
      status: "stopped",
      runningCount: request.runningCount,
      waitingApprovalCount: request.waitingApprovalCount,
      message: "iOS uses resume backfill instead of continuous background sync"
    )
    emit(snapshot)
    finishCleanupTask(message: "iOS background cleanup completed")
    return snapshot
  }

  private func finishCleanupTask(message: String) {
    endBackgroundTaskIfNeeded()
    emit(stoppedSnapshot(message: message))
  }

  private func endBackgroundTaskIfNeeded() {
    guard backgroundTask != .invalid else { return }
    application?.endBackgroundTask(backgroundTask)
    backgroundTask = .invalid
  }

  private func stoppedSnapshot(message: String) -> [String: Any?] {
    makeSnapshot(
      status: "stopped",
      runningCount: 0,
      waitingApprovalCount: 0,
      message: message
    )
  }

  private func emit(_ snapshot: [String: Any?]) {
    if snapshot["status"] as? String == "stopped" {
      lastSnapshot = nil
    } else {
      lastSnapshot = snapshot
    }
    eventSink?(snapshot)
  }

  private func makeSnapshot(
    status: String,
    runningCount: Int,
    waitingApprovalCount: Int,
    message: String?
  ) -> [String: Any?] {
    [
      "status": status,
      "runningCount": runningCount,
      "waitingApprovalCount": waitingApprovalCount,
      "message": message,
    ]
  }

  private func request(from args: [String: Any]?) -> BackgroundConversationSyncRequest? {
    guard let args else { return nil }
    return BackgroundConversationSyncRequest(
      runningCount: args["runningCount"] as? Int ?? 0,
      waitingApprovalCount: args["waitingApprovalCount"] as? Int ?? 0
    )
  }
}

private struct BackgroundConversationSyncRequest {
  let runningCount: Int
  let waitingApprovalCount: Int
}

File: mobile/android/app/src/main/kotlin/com/example/lan_ai_cli_control/BackgroundConversationSyncChannels.kt
package com.example.lan_ai_cli_control

import android.content.Context
import androidx.annotation.MainThread
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel

object BackgroundConversationSyncChannels {
    private const val METHOD_CHANNEL = "lan_ai_cli_control/background_conversation_sync"
    private const val EVENT_CHANNEL = "lan_ai_cli_control/background_conversation_sync/events"
    private var eventSink: EventChannel.EventSink? = null
    private var lastSnapshot: Map<String, Any?>? = null

    fun register(context: Context, flutterEngine: FlutterEngine) {
        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, METHOD_CHANNEL)
            .setMethodCallHandler { call, result ->
                when (call.method) {
                    "isSupported" -> result.success(true)
                    "start" -> {
                        val args = call.arguments as? Map<*, *>
                        val request = requestFromMap(args)
                        if (request == null) {
                            result.error(
                                "BAD_ARGUMENT",
                                "Invalid background conversation sync request",
                                null
                            )
                            return@setMethodCallHandler
                        }
                        try {
                            BackgroundConversationSyncService.start(
                                context.applicationContext,
                                request
                            )
                            val snapshot = snapshot(
                                "starting",
                                request.runningCount,
                                request.waitingApprovalCount,
                                "Starting background sync"
                            )
                            emit(snapshot)
                            result.success(snapshot)
                        } catch (error: Exception) {
                            val snapshot = snapshot(
                                "denied",
                                request.runningCount,
                                request.waitingApprovalCount,
                                error.message ?: "Android rejected background sync"
                            )
                            emit(snapshot)
                            result.success(snapshot)
                        }
                    }
                    "stop" -> {
                        BackgroundConversationSyncService.stop(context.applicationContext)
                        emit(snapshot("stopped", 0, 0, "Background sync stopped"))
                        result.success(null)
                    }
                    "snapshot" -> result.success(lastSnapshot)
                    else -> result.notImplemented()
                }
            }

        var ownedSink: EventChannel.EventSink? = null
        EventChannel(flutterEngine.dartExecutor.binaryMessenger, EVENT_CHANNEL)
            .setStreamHandler(object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                    ownedSink = events
                    eventSink = events
                    lastSnapshot?.let { events?.success(it) }
                }

                override fun onCancel(arguments: Any?) {
                    if (eventSink === ownedSink) {
                        eventSink = null
                    }
                    ownedSink = null
                }
            })
    }

    @MainThread
    fun emit(snapshot: Map<String, Any?>) {
        val status = snapshot["status"] as? String
        if (status == "stopped") {
            lastSnapshot = null
        } else {
            lastSnapshot = snapshot
        }
        eventSink?.success(snapshot)
    }

    fun snapshot(
        status: String,
        runningCount: Int,
        waitingApprovalCount: Int,
        message: String?
    ): Map<String, Any?> = mapOf(
        "status" to status,
        "runningCount" to runningCount,
        "waitingApprovalCount" to waitingApprovalCount,
        "message" to message
    )

    private fun requestFromMap(args: Map<*, *>?): BackgroundConversationSyncRequest? {
        if (args == null) return null
        return BackgroundConversationSyncRequest(
            runningCount = (args["runningCount"] as? Number)?.toInt() ?: 0,
            waitingApprovalCount = (args["waitingApprovalCount"] as? Number)?.toInt() ?: 0,
            notificationTitle = args["notificationTitle"] as? String ?: "Vibe Coding",
            notificationBody = args["notificationBody"] as? String ?: "Background sync active"
        )
    }
}

data class BackgroundConversationSyncRequest(
    val runningCount: Int,
    val waitingApprovalCount: Int,
    val notificationTitle: String,
    val notificationBody: String
)
